Make public "Find User" functionality configurable

christoph-heiss opened the pull request:
Hi!

Really like Note Mark so far!

One thing I noticed - it would probably be desirable to allow users to disable the public-accesible "Find User" functionality on the homepage and restrict it to authenticated users only.

I've implemented it using a config env flag on the backend and passing it to the frontend via the `/info` API endpoint.

Thanks in advance for consideration!

enchant97 commented:
Hi thanks for your interest in the project, what benefit do you think there is to disabling this?

christoph-heiss commented:
> what benefit do you think there is to disabling this?

So basically just _not_ exposing possible usernames - this features makes them pretty easy to enumerate.

So one big part is simply security - not exposing more info then needed -.
But also privacy is a good reason. Esp. if you share your instances with e.g. family/friends/etc, you might not want people to easily find out all their (user)names. 

So having a search user functionality as authenicated user is pretty useful, but you might not want it to be accessible to the world.

christoph-heiss commented on backend/config/config.go:
```diff
@@ -17,14 +17,15 @@ type DBConfig struct {
+	StaticPath         string        `env:"STATIC_PATH"`
+	CORSOrigins        []string      `env:"CORS_ORIGINS,notEmpty" envSeparator:","`
+	AllowSignup        bool          `env:"ALLOW_SIGNUP,notEmpty" envDefault:"true"`
+	AnonymousFindUser  bool          `env:"ANONYMOUS_FIND_USER,notEmpty" envDefault:"true"`
```
Sounds good! Naming is always hard and I wasn't sure if a rather long name would be acceptable.

christoph-heiss commented on frontend/src/routes/(home).tsx:
```diff
@@ -23,6 +24,17 @@ const Home: Component = () => {
     })
   }
 
+  const anonymousFindUserAllowed = createAsync(async () => {
```
Yeah, I've tried that, but at this point the API/server-info has not been queried.

But I'll look over it again and implement it properly - also haven't worked with solid-js before, so I'm probably just missing something.

christoph-heiss commented:
I have now fixed all the above suggestions now. 

But, I changed it so that the button is also hidden if the API cannot be reached (e.g. invalid url). IMO doesn't make sense to show it anyway it that situation, as it (obviously) won't work. But happy to change it back if you prefer.

enchant97 commented on frontend/src/contexts/ApiProvider.tsx:
```diff
@@ -46,6 +46,13 @@ const makeApiContext = () => {
   createEffect(() => {
     writeApiDetails(apiConfig())
+
+    api().getServerInfo()
```
This should not be needed as info is gathered on initial request to api and stored for the current session. It may however be `undefined` whilst it is being gathered or no valid api server is selected.

christoph-heiss replied:
Unfortunately, that is not the case. `getServerInfo()` never gets called on the home page, for unauthenticated session, at least.
That's what had me confused at first. Can be seen pretty easily just by opening the dev tools before opening the page. :sweat_smile:

enchant97 replied:
Thanks for finding a bug! Instead of your fix, could you implement the below in `frontend/src/index.tsx`:

```diff
 render(() => (
   <Router>
     <Route path="/" component={Wrapper}>
-      <Route path="/" component={Home} />
+      <Route path="/" component={() => <RequireApiSetupGuard><Home /></RequireApiSetupGuard>} />
       <Route path="/scratch-pad" component={ScratchPad} />
       <Route component={RequireNoAuthGuard}>
         <Route path="/signup" component={() => <RequireSignupAllowedGuard><Signup /></RequireSignupAllowedGuard>} />
```

This should ensure info is requested correctly.

christoph-heiss replied:
Sure! Thanks for looking into it, looks like a lot cleaner solution!

christoph-heiss replied:
Just tried it out - a side-effect of this is that the user now get's directly redirected to the login page if the server is unreachable, due to [this line](https://github.com/enchant97/note-mark/blob/8d499a791653da7d2de41ece9d90b2cc6e94cc1f/frontend/src/contexts/ApiProvider.tsx#L104).

Thinking about it, it makes sense tho - if the API is unreachable, the only really useful thing to do is setting the API URL anyway, so redirecting to the login page isn't wrong IMO.

enchant97 reviewed (approved):
Looks good. Thanks for your work :)

enchant97 commented:
Your feature has now been integrated into the project! It will appear in the next feature update (also with the new backend).
